### ISSUE TITLE
Can't get mangle with properties to output usable code

### ISSUE THREAD

**Issue by Friksel** [terser-js/terser#196]
Using terser via terser-webpack-plugin I'm trying for days now to output working code with mange.properties set. But without any success. Eventhough terser should have everything in one javascript file as compiled by webpack 4, there's always stuff it doesn't recognise.

Diving deeper into it I see now the mangler even mangles DOM-properties and functions. Even things like 'body' and 'getAttribute' are being mangled. It doesn't matter if I put toplevel to true or false btw.

After days of trying pretty much every setting there is to the mangler, and posting questions on other forums without any reaction, I hope you can help me. Because I really start to wonder how people ever get mangle to work with properties.

I tried setting the regex in mangle.properties to the only object I'd like to mangle. But it seems to just replace all words it can find like that, so that seems pretty errorprone if I don't use uniquely named properties.

Also excluded the used libs (like gsap and createjs) to outside webpack (just for testing), because most problems with mangle.properties seem to arrise on the libs. But in the end it seems to be more the cause of the problem, because the libs are in the beginning of the webpack output file.

I searched through all settings of Terser and tried a lot. Also looked all over the internet for people really using mangle on terser or uglifyjs, which seem to have the (almost) exact API interface. 

But it's almost as if nobody really uses the mangler. Or is pretty quiet about it.
I find it hard to believe nobody uses mangle, but can't find anything about it either. I also would expect it to work out of the box on a javascript file with everything inside.

Could anybody here please help me and tell what could cause a mangle to not result in a useful file on javascript code which has everything inside in one file? Is there something I should know about or am missing here?
And is there perhaps some documentation or blog or something else on this mangle properties that can help me (and perhaps other) to really use mangle.properties in a project? (next to the docs which are pretty summier on this subject, although it would be enough if it would work directly out of the box... but that doesn't seem to be the case).

Sorry for the long text and I understand this is not the place to ask questions, but there seem to be knowbody having an answer to my question on forums, so I came back to the source in hope for help!

Thanks!

**Comment by emilemil1** [comment]
I have the same question. I'd really like to mangle properties in case I have something like:

`let exampleObject = {longAndVerbosePropertyNameA: 1, longAndVerbosePropertyNameB: 2}`

But as Friksel says, enabling property mangling will also completely destroy any properties of DOM objects such as window or document. There is a whitelist file (tools/domprops.json), but only the CLI seems to be able to use it. It even has a `--mangle-props domprops` flag to disable the whitelist.

What I've had to do to get it working is to load this whitelist manually and apply it using the reserved option. Then I had to append a list of every possible externally defined property, most notably anything sent or received as JSON. It'd be great if terser could at least detect which objects are declared externally and not mangle those. Anyway, after doing this and fixing one small introduced code error, everything worked fine.

As a feature suggestion, a slightly unsafe alternative to mangling externals would be to reassign them. For example: `let a = document` It will basically do the same thing as mangling, at the minimal cost of allocating an object reference. The same thing can then be done for properties as well, so you could do: `a.b = a.getAttribute` The only things this breaks are key enumeration/iteration and sending the objects as JSON. Whitelists are decent enough for those cases. It's still better than having no compression.

**Comment by fabiosantoscode** [comment]
This is not what I'm seeing. Can you share a code sample with a terser command or API call? Using only `terser [file] --mangle-props` it works for me.

**Comment by emilemil1** [comment]
I've only used the npm module, not the command line. Here's an example:

testScript.js:
```
const terser = require("terser");
const fs = require("fs");
let text = fs.readFileSync("test.js").toString();
let settings = {
    "mangle": {
        "properties": true
    }
}
console.log(terser.minify(text, settings).code);
```

test.js:
`document.createElement('div')`

output:
`document.d("div");`

`createElement` is being mangled to `d` even though it shouldn't.

**Comment by Friksel** [comment]
I'm having comparable results, also with npm (although using via another layer on top: webpack-terser-plugin)

**Comment by fabiosantoscode** [comment]
It's been fixed now. Will release soon.

**Comment by fabiosantoscode** [comment]
If I fixed your error consider contributing on patreon ;)

**Comment by Friksel** [comment]
Thanks @fabiosantoscode !

**Comment by Friksel** [comment]
Today I had a little time to do some testing, and it works much better now. But for some reason after mangling a single closed code file, there are still things the browser can't find, which are not easy to debug:

![mangle-problem](https://user-images.githubusercontent.com/3089978/50382414-c61e8880-069e-11e9-8676-bbd767a5ed75.JPG)

As you see it's looking for some Or() method inside the s-function which is a self-executing function that just got created. It looks like the s-function should return some constructor function which should have some Or()-method, but like this it's pretty hard to follow what's going wrong here.
btw this part of the code is from a lib I use. Not sure which one, but could be pixi.js or greensock gsap.

I hope to find some more time soon to dive in deeper into this and hopefully to wrap my head around this. But I'm open to clues cause I would expect mangling to work on a single file where all references are known inside because all code is there. So if not working like that, to me that will be gibberish feeling pretty soon I'm afraid...

**Comment by emilemil1** [comment]
This is what I can gather from that snippet. The self-executing function clones an object, and it also sets the clone's .default property to be the original object. Or() is supposed to be some constructor of the original object.

What is Or() supposed to be named in the original code? Assuming that it's a constructor, where is that constructor defined, and has it also been mangled to Or? If it has, then the cloning might have broken. If not, where does Or() come from, what other properties have been assigned that name? Has the constructor name been used multiple times for different things so they are getting mixed up?

**Comment by fabiosantoscode** [comment]
We need more info. @Friksel can you try to minimize that sample as small as possible, and paste it as text so we can test it?

Happy holidays from me and Donald J Trump!

**Comment by Friksel** [comment]
@fabiosantoscode  What you see is a bundle of my code and imported libs bundled by webpack. The fragment is all inside the pixi.js lib. So I can't make it smaller than this. PixiJs is here: https://github.com/pixijs/pixi.js. Hope this helps

**Comment by Friksel** [comment]
@fabiosantoscode To this date unfortunetely it I couldn't get mangle to work with any webpack/babel project here. I've tested a lot of different projects. I can't get mangle to work even on projects without any libs being used, so only my code. Common factor though is that I use webpack 4 and babel on each project, so after a lot of testing I can only come to the conclusion Terser somehow at least can't handle the webpack generated output to mangle or I am missing a terser setting to make mangle work. I walked through all terser-settings, but can't find anything wrong so far. But I might be missing something.

Today I created a stripped webpack/babel project with only the very basics in it. In src/js you'll find only a simple es6 class `Main` with just a constructor and a simple function in it. That's all there is in this project to mangle. In the attached projects.zip file I include this sample project (install with `npm install`) and I also attach a zip with two already build dist results; one with mangle turned on and one with mangle turned off.

As you can see the result build with mangle off is working, but the one with mangle on is not, because it can't find things. It looks this has to do with the way webpack/babel output functions, but because it's still valid javascript Terser should be able to mangle this. 
If you like to change and build the project, just run script `npm run dist` to 'transpile' to a new dist in the project. The used terseroptions are in the root of the project.

[dist-results.zip](https://github.com/terser-js/terser/files/2754666/dist-results.zip)
[project.zip](https://github.com/terser-js/terser/files/2754667/project.zip)

Hope this helps. It would be very welcome if mangling works.

**Comment by fabiosantoscode** [comment]
This is to be expected as not all projects are going to work, so there's a warning in the docs.

The question still stands, can you send a minimal code for reproducing this? It can be done in 10 lines or less..

**Comment by Friksel** [comment]
> The question still stands, can you send a minimal code for reproducing this? It can be done in 10 lines or less..

@fabiosantoscode I just sent you. I wonder if you even looked at the zips. If you opened the zip file you you would have seen this is the only js source in there:

```
export default class Main {

    constructor() {
        this.bEl = [];
        this.createBlocks();
    }

    createBlocks() {
        for (let i = 0; i < 3; i++) {
            this.bEl[i] = document.createElement('div');
            this.bEl[i].classList.add('block');
            document.body.appendChild(this.bEl[i]);
        }
    }
}

```

As said this got transpiled by webpack/babel to the dist files I send you too that pass through Terser. 
There are no libs used, no external sources, nothing complicated here. Just one file with legal javascript and everything inside.

**Comment by songololo** [comment]
I am no expert on Webpack or Terser but I'm wondering about two things that might reduce the occurrence of some of the above issues:

- Is there a simple way to only mangle specific files? (i.e. without turning off minification for other files).
- Is there a way to mangle property names before babel does its thing?

**Comment by fabiosantoscode** [comment]
Reopening this as indeed I didn't have a look at the zips

**Comment by fabiosantoscode** [comment]
@songololo you can do 1. in your module bundler. In webpack you can use a test regex or function to select a few files, and minify those with certain options, and select the other files and minify with other options. This can be used for example to not minify some libraries that you know are already small.

For 2. you can do it if you're using pure javascript, even if it's ES6 or above (eg. no `<JSX />` tags). Terser supports it.

**Comment by fabiosantoscode** [comment]
This is one of the 2 bugs I'm looking at this weekend.

**Comment by fabiosantoscode** [comment]
First off, I've changed the warning about mangle with properties breaking your code from "might break your code" to "will break your code", since it's more truthful. Mangling properties is impossible to do correctly unless all the code you're using is either very simple, uses prefixed properties, and of course Object.defineProperty and friends (functions that get or manipulate object properties by name) won't work correctly, since terser won't know to mangle the strings being passed into it.

I've run your class through Terser and I got something useful back:

```
// (it works with the class you sent too, this is just smaller output)
function x() {
    for (let i = 0; i < 3; i++) {
        this.bEl[i] = document.createElement('div');
        this.bEl[i].classList.add('block');
        document.body.appendChild(this.bEl[i]);
    }
}

foo(x)
```

This resulted in a piece of code that has the words "classList", "body", etc. as expected.

In your the screenshot you sent `s` seems to be an object, as on top it returns the input object (we can be pretty sure it's an object because the code is checking for the `.l` property).

The problem you're having in that code, I believe, is with the hasOwnProperty call. It receives a string, but that string isn't mangled by Terser, so the object copy that's happening isn't being completed successfully.
